### PR TITLE
Add support for CLI application code generation

### PR DESCRIPTION
**Pull request by falzm**
This change introduces minimalist but fully functional CLI application code generation based on an OpenAPI spec. Taking the example Petstore OpenAPI spec from your project, this is the result:

```console
$ go run examples/petstore-expanded/cli/main.gen.go --help
NAME:
   cli - A new cli application

USAGE:
   cli [global options] command [command options] [arguments...]

COMMANDS:
   find-pets       Returns all pets
   add-pet         Creates a new pet
   delete-pet      Deletes a pet by ID
   find-pet-by-id  Returns a pet by ID
   help, h         Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --server value, -S value  server to send requests to
   --output value, -O value  output format (text|table|json) (default: "text")
   --help, -h                show help (default: false)

$ go run examples/petstore-expanded/cli/main.gen.go add-pet --help
NAME:
   cli add-pet - Creates a new pet

USAGE:
   cli add-pet [command options] [arguments...]

OPTIONS:
   --body value  request body
   --help, -h    show help (default: false)

$ go run examples/petstore-expanded/cli/main.gen.go delete-pet --help
NAME:
   cli delete-pet - Deletes a pet by ID

USAGE:
   cli delete-pet [command options] [arguments...]

OPTIONS:
   --id value  ID of pet to delete (default: 0)
   --help, -h  show help (default: false)
```

When an API operation request returns data in the response, the CLI supports several output format (example from an internal generated CLI):

```console
$ cli -O text list-files
299eaecf83e5a05141c5b1731b14d974    2020-02-28 16:58:24.238201015 +0100 CET 287b6306-fdeb-4dc6-855d-90c4f68f572b    /tmp/test/files/287b6306-fdeb-4dc6-855d-90c4f68f572b
b2d1236c286a3c0704224fe4105eca49    2020-02-21 12:18:40.669464419 +0100 CET 447f5efe-fe73-4ab0-8ac3-dfeecb5d988d    /tmp/test/files/447f5efe-fe73-4ab0-8ac3-dfeecb5d988d
b6d81b360a5672d80c27430f39153e2c    2020-02-21 12:18:40.67124133 +0100 CET  5e12e10a-c0ac-4ed4-895a-cc49b4ad4386    /tmp/test/files/5e12e10a-c0ac-4ed4-895a-cc49b4ad4386

$ cli -O table list-files
+----------------------------------+--------------------------------+--------------------------------------+------------------------------------------------------+
|             CHECKSUM             |              DATE              |                  ID                  |                         PATH                         |
+----------------------------------+--------------------------------+--------------------------------------+------------------------------------------------------+
| 299eaecf83e5a05141c5b1731b14d974 | 2020-02-28 16:58:24.238201015  | 287b6306-fdeb-4dc6-855d-90c4f68f572b | /tmp/test/files/287b6306-fdeb-4dc6-855d-90c4f68f572b |
|                                  | +0100 CET                      |                                      |                                                      |
| b2d1236c286a3c0704224fe4105eca49 | 2020-02-21 12:18:40.669464419  | 447f5efe-fe73-4ab0-8ac3-dfeecb5d988d | /tmp/test/files/447f5efe-fe73-4ab0-8ac3-dfeecb5d988d |
|                                  | +0100 CET                      |                                      |                                                      |
| b6d81b360a5672d80c27430f39153e2c | 2020-02-21 12:18:40.67124133   | 5e12e10a-c0ac-4ed4-895a-cc49b4ad4386 | /tmp/test/files/5e12e10a-c0ac-4ed4-895a-cc49b4ad4386 |
|                                  | +0100 CET                      |                                      |                                                      |
+----------------------------------+--------------------------------+--------------------------------------+------------------------------------------------------+

$ cli -O json list-files | jq .
[
  {
    "checksum": "299eaecf83e5a05141c5b1731b14d974",
    "date": "2020-02-28T16:58:24.238201015+01:00",
    "id": "287b6306-fdeb-4dc6-855d-90c4f68f572b",
    "path": "/tmp/test/files/287b6306-fdeb-4dc6-855d-90c4f68f572b"
  },
  {
    "checksum": "b2d1236c286a3c0704224fe4105eca49",
    "date": "2020-02-21T12:18:40.669464419+01:00",
    "id": "447f5efe-fe73-4ab0-8ac3-dfeecb5d988d",
    "path": "/tmp/test/files/447f5efe-fe73-4ab0-8ac3-dfeecb5d988d"
  },
  {
    "checksum": "b6d81b360a5672d80c27430f39153e2c",
    "date": "2020-02-21T12:18:40.67124133+01:00",
    "id": "5e12e10a-c0ac-4ed4-895a-cc49b4ad4386",
    "path": "/tmp/test/files/5e12e10a-c0ac-4ed4-895a-cc49b4ad4386"
  }
]
```

This PR brings a significant feature and quite a chunk of code, so naturally I expect change requests from you – and you can of course refuse it plainly, I'd understand. Note that at this point I don't consider the PR finished (some tests and possibly a quick blurb in the README would be a nice addition IMO), but I preferred to open it anyway so you can challenge the implementation now rather than later.  Here are some points on which I'd like to have a feedback:

* Contrary to the other targets, this new one generates a fully functional CLI app – not just boiler plate code. This is heavily inspired from [openapi-cli-generator](https://github.com/danielgtaylor/openapi-cli-generator), which is a very nice project but only generates client-side code and doesn't seem to be *actively* maintained [despite what its author says](https://github.com/danielgtaylor/openapi-cli-generator/issues/26). The rationale behind this PR is to have a turn-key CLI for free with your OpenAPI spec that is actually nice from a UX perspective – although it won't ever beat a manually crafted tool in this regard – but I'd understand if you'd prefer to stay consistent with the server-side generation targets and also limit this feature to command stubs, and let the user figure out the rest of the CLI.
* I've used [spf13/cobra](https://github.com/spf13/cobra) as CLI framework because it's one of the most popular ones in the Go community, but I don't have particular preference for it so that could very well change.
* The requests sending a body proved to be challenging to implement with the current architecture of the code generation framework. I've had to settle for a generic `--body="<some JSON string>"` form because despite my efforts I couldn't manage to access the internal body object elements in order to translate them to individual CLI command parameters as I did for query/path parameters, and in true 80/20 principle I wasn't willing to spend more time on this because it's not in our current internal use case.

The current implementation is quite tailor-made for our internal use case, but I'm more than willing to make it more generic so it can fit others – up to a certain point where it'll make more sense for us to fork the project altogether 😉

### CODE REVIEW

**Comment by deepmap-marcinr** [comment]
Whoa. This is a big change, but also very cool.

One major comment is that rather than duplicating all the client code into your cli/ directory, you should just import the existing client, and implement the CLI in terms of that.

**Comment by falzm** [comment]
Happy to hear that you find this feature useful :)

I'm not sure to understand what you mean by "import the existing client, and implement the CLI in terms of that", as the `cli-app` effectively relies on the other`types`/`client` targets output (see [here](https://github.com/deepmap/oapi-codegen/pull/144/files#diff-02c777a104e63c30ccc0115c8a78c90aR17)): can you elaborate?

**Comment by aca** [comment]
@falzm this would really be great and useful!

**Comment by falzm** [comment]
Update: due to various implementation quirks and [the overall state of the Cobra project](https://github.com/spf13/cobra/issues/959) I've eventually switched to [urfave/cli.v2](https://github.com/urfave/cli) as the CLI framework, which is also very popular and seems better maintained than the Cobra.

Also, I've implemented a way for users to customize the generated client options and CLI startup hooks: for example to inject a *correlation ID* to every request sent by the client, this snippet can be added to a file located in the same package as the generated client code:

```go
func init() {
	// clientOptions is a slice of ClientOption functions passed to NewClient()
	clientOptions = append(clientOptions, WithRequestEditorFn(injectCID))

	// appbeforeFuncs is a slice of cli.BeforeFunc executed in the app.Before function at CLI startup
	appBeforeFuncs = append(appBeforeFuncs, loadCIDToCLIContext)

	app.Flags = append(app.Flags, &cli.StringFlag{
		Name:  "cid",
		Usage: "correlation ID",
	})
}

// injectCID extracts a correlation ID from the global context and injects it in a HTTP header of the request.
// If no CID value is found in the context, a new one is generated.
func injectCID(req *http.Request, ctx context.Context) error { // nolint:golint
	var cid string

	if v, ok := ctx.Value("cid").(string); ok && v != "" {
		cid = v
	} else {
		cid = xid.New().String()
	}

	req.Header.Add("X-CID", cid)

	return nil
}

// loadCIDToCLIContext looks up the CLI application flags and checks if a value has been provided for the "--cid" flag,
// and loads it as CLI context.Context value if one is found.
func loadCIDToCLIContext(ctx *cli.Context) error {
	if cid := ctx.String("cid"); cid != "" {
		ctx.Context = context.WithValue(ctx.Context, "cid", cid)
	}

	return nil
}
```

File: cmd/oapi-codegen/oapi-codegen.go [hunk 1]
```diff
@@ -69,6 +69,8 @@ func main() {
 		switch g {
 		case "client":
 			opts.GenerateClient = true
+		case "cli-app":
+			opts.GenerateCLIApp = true
```
[the remark this answers is not in the document]
Comment (falzm) [reply cmd/oapi-codegen/oapi-codegen.go]: That's a good point, however I'd suggest leaving it to the user's discretion to decide which part of the code is output to which file, e.g. personally I like to split the code in individual files per "component" (`types.gen.go`, `client.gen.go` etc) as it makes searching/reading easier.

File: pkg/codegen/template_helpers.go [hunk 2]
```diff
@@ -85,6 +85,69 @@ func genParamFmtString(path string) string {
 	return ReplacePathParamsWithStr(path)
 }
 
+// This function takes an array of Parameter definition, and generates a map of
+// key/value pairs with parameter names as key and parameter type definition as
+// value.
+func genParamsMap(params []ParameterDefinition) map[string]string {
```
[the remark this answers is not in the document]
Comment (falzm) [reply pkg/codegen/template_helpers.go]: Changed as suggested, thanks.

**Comment by deepmap-marcinr** [comment]
So, one overall question. It appears that you can specify base types and string arrays via the cli - are you planning on doing something more complex, like POSTing an object with multiple properties? I wonder if it's possible to do something like HTTPie does or maybe even like deepObject based params. "cli postSomething obj[field1]=x obj[field2]=y". It's an annoying problem to wrap hierarchical data in a flat cli.

This does look cool, though!

**Comment by falzm** [comment]
Well as I've explained in it the PR description: at this point I'm not sure how to handle complex objects/parameters scenario, as we don't have this use case at the moment. My suggestions is to leave it *as is* for the time being, and improve it if/when someone comes up with an actual use for those.

**Comment by deepmap-marcinr** [comment]
I wasn't asking for changes, missed that line in your description. I'm ok
with this going in as-is, with a documented caveat that it's limited. Once
you're ready, I'll merge it.

**Comment by falzm** [comment]
Great, thanks!

**Comment by falzm** [comment]
Hi, we're pulling the plug on this PR – we've decided to go with another route internally, eventually. Thank you for showing interest in this proposal anyway :)

**Pull request closed without merging**